pkg/cli/server.go

```diff
@@ -239,19 +245,15 @@ func (svc *serverService) stop() {
 	svc.selfProfiling.Stop()
 	svc.logger.Debug("stopping upstream")
 	svc.directUpstream.Stop()
-	svc.logger.Debug("stopping storage")
-	if err := svc.storage.Close(); err != nil {
-		svc.logger.WithError(err).Error("storage close")
-	}
 	svc.logger.Debug("stopping http server")
 	if err := svc.controller.Stop(); err != nil {
 		svc.logger.WithError(err).Error("controller stop")
 	}
-	if svc.config.EnableExperimentalAdmin {
-		svc.logger.Debug("stopping admin server")
-		if err := svc.adminServer.Stop(); err != nil {
-			svc.logger.WithError(err).Error("admin server stop")
-		}
+
+	// storage needs to be the last thing to shutdown
+	svc.logger.Debug("stopping storage")
+	if err := svc.storage.Close(); err != nil {
+		svc.logger.WithError(err).Error("storage close")
```

Comment:
I remembered why `controller` is stopped in the last turn:
1. We may want to observe metric values while storage is closing.
2. I think `/healthz` endpoint (which is used as liveness probe) should be still responding in order let storage flush everything to disk before server gets killed (although, I'm not 100% sure: e.g in kubernetes there is `terminationGracePeriodSeconds` option for this).

Reply:
interesting points, for 2) I wonder if we shouldn't leverage the `preStop` hook instead, so we have more confidence that our shutdown is going to finish

but for now I've undone the order change